### PR TITLE
Fix Mimi Bug

### PR DESCRIPTION
**Pull request by poonehmousavi**
## What does this PR do?

- Mimi does not function correctly on CUDA because the embeddings are always loaded onto the CPU (in the __init__ function) and are never transferred to CUDA. To address this, I followed the approach in the original code by calling _compute_embedding within the forward or encode function on the first call to ensure the embeddings are loaded onto the correct device.

- Added a sampling_rate attribute for both WavTokenizer and SpeechTokenizer.

-->

Fixes #<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/lobes/models/huggingface_transformers/mimi.py [hunk 1]
```diff
@@ -115,6 +115,9 @@ def forward(self, inputs, length):
         audio : torch.Tensor
             the reconstructed audio
         """
+        if self.embeddings is None:
+            self.embeddings = self._compute_embedding()
+
```
Comment (pplantinga) [inline speechbrain/lobes/models/huggingface_transformers/mimi.py:120]: Not sure this is needed since `forward` directly calls `encode`

File: speechbrain/lobes/models/huggingface_transformers/mimi.py [hunk 2]
```diff
@@ -178,6 +184,9 @@ def decode(self, tokens, length=None):
...
```
Comment (pplantinga) [inline speechbrain/lobes/models/huggingface_transformers/mimi.py:189]: Not sure if this is needed, since the embeddings don't seem to be used later in `decode`

**Review by pplantinga — COMMENTED** [review]
Looks pretty good, just a few questions to answer before mreging

File: speechbrain/lobes/models/huggingface_transformers/mimi.py [hunk 3]
```diff
@@ -78,8 +78,8 @@ def __init__(
 
         super().__init__(source=source, save_path=save_path, freeze=freeze)
         self.num_codebooks = num_codebooks
-        self.sampling_rate = sample_rate
-        self.embeddings = self._compute_embedding()
+        self.sample_rate = sample_rate
+        self.embeddings = None
```
[the remark this answers is not in the document]
Comment (poonehmousavi) [reply speechbrain/lobes/models/huggingface_transformers/mimi.py:82]: Apologies if my explanation wasn't clear earlier. I added self.embedding, which caused this [function](https://github.com/huggingface/transformers/blob/6bc0fbcfa7acb6ac4937e7456a76c2f7975fefec/src/transformers/models/mimi/modeling_mimi.py#L1261) to be called during initialization. As a result, self._embed becomes non-None.

In the original Mimi code, this function is first invoked when quantize is called, ensuring that self._embed is on the correct device. However, because I called the [_compute_embedding](https://github.com/poonehmousavi/speechbrain/blob/e5796d2d6a2a2be725520b83b7ac2130b32ae14d/speechbrain/lobes/models/huggingface_transformers/mimi.py#L85) function in the __init__ method, it leads to a "not on the same device" error.

To resolve this, I followed a similar approach and called _compute_embedding in the encode function instead. I hope this provides more clarity. Based on this explanation, what would you suggest?

File: speechbrain/lobes/models/huggingface_transformers/mimi.py (hunk 1 above)
Comment (poonehmousavi) [reply speechbrain/lobes/models/huggingface_transformers/mimi.py:120]: You are right, i could remove it.

File: speechbrain/lobes/models/huggingface_transformers/mimi.py (hunk 2 above)
Comment (poonehmousavi) [reply speechbrain/lobes/models/huggingface_transformers/mimi.py:189]: iIn the original code, the decode function calls the [self.emb](https://github.com/huggingface/transformers/blob/6bc0fbcfa7acb6ac4937e7456a76c2f7975fefec/src/transformers/models/mimi/modeling_mimi.py#L1285) function. This could potentially lead to the same "not on the same device" issue if decode is called before encode.

Such a scenario is possible when only the decoder module is used to convert tokens to signals, bypassing the encode function entirely

File: speechbrain/lobes/models/discrete/speechtokenizer.py [hunk 4]
```diff
@@ -37,6 +37,8 @@ class SpeechTokenizer(nn.Module):
         HuggingFace hub name: e.g "fnlp/SpeechTokenizer"
     save_path : str
         Path (dir) of the downloaded model.
+    sample_rate : int (default: 16000)
+        The audio sampling rate
```
[the remark this answers is not in the document]
Comment (poonehmousavi) [reply speechbrain/lobes/models/discrete/speechtokenizer.py:41]: Most of these codecs are used in [DASB benchmark](https://github.com/speechbrain/benchmarks/tree/DASB/benchmarks/DASB), we always check sampling rate of the tokenizers and up/dowsample the audio according to this sampling rate. Example : [here](https://github.com/speechbrain/benchmarks/blob/9f64966ca6f0467dc140b7047ac2d908038fefc1/benchmarks/DASB/utils/tokens.py#L194)

[the remark this answers is not in the document]
Comment (pplantinga) [reply speechbrain/lobes/models/discrete/speechtokenizer.py:41]: Aha, this makes sense!

File: speechbrain/lobes/models/huggingface_transformers/mimi.py (hunk 2 above)
Comment (pplantinga) [reply speechbrain/lobes/models/huggingface_transformers/mimi.py:187]: Aha, makes sense!

**Review by pplantinga — APPROVED** [review]
I tried this, it works, so LGTM!
